### PR TITLE
Add lattice classes

### PR DESCRIPTION
**Pull request by michaelpj**
This adds lattice classes to `plutus-tx`. We use them mainly for
`Interval`, but also in a few other places.

It's a little awkward putting them in `plutus-tx` when we're primarily
using them off-chain, but:
- We probably do want to use them on-chain when working with
`Interval`s.
- When we want to use a class on-chain we have to duplicate it if it's
from a non-controlled source.

So given that we'll need a copy in `plutus-tx`, it seems like we might
as well just stick with the one copy for now.

This lets us sensibly define a monoid instance for `Tx` too, which is
nice.

There is a hack in the second commit to make `deriving Schema` compile. This is wrong, I need to work with @krisajenkins to fix it, but I wanted to check that the direction looked good to people first.

### CODE REVIEW

File: plutus-wallet-api/ledger/Ledger/Slot.hs [hunk 1]
```diff
@@ -44,15 +42,3 @@ makeLift ''Slot
 
 -- | An 'Interval' of 'Slot's.
 type SlotRange = Interval Slot
-
--- | A 'SlotRange' that covers only a single slot.
-singleton :: Slot -> SlotRange
-singleton (Slot s) = Interval (Just (Slot s)) (Just (Slot (plus s 1)))
-
--- | Number of 'Slot's covered by the interval. @width (from x) == Nothing@.
-width :: SlotRange -> Maybe Integer
```
Comment (j-mueller) [inline plutus-wallet-api/ledger/Ledger/Slot.hs]: Where did that go?

File: plutus-wallet-api/ledger/Ledger/Tx.hs [hunk 2]
```diff
@@ -140,6 +142,19 @@ data Tx = Tx {
     } deriving stock (Show, Eq, Generic)
       deriving anyclass (ToJSON, FromJSON, Serialise)
 
+instance Semigroup Tx where
```
Comment (j-mueller) [inline plutus-wallet-api/ledger/Ledger/Tx.hs]: we also need an instance for `Language.Plutus.Contract.Tx.UnbalancedTx`

File: plutus-wallet-api/ledger/Ledger/Slot.hs (hunk 1 above)
Comment (michaelpj) [reply plutus-wallet-api/ledger/Ledger/Slot.hs]: Oops.

File: plutus-wallet-api/ledger/Ledger/Tx.hs (hunk 2 above)
Comment (michaelpj) [reply plutus-wallet-api/ledger/Ledger/Tx.hs]: :+1:

**Comment by michaelpj** [comment]
Note: Kris is on holiday, so this probably isn't going anywhere until he gets back, but that's okay.

**Comment by michaelpj** [comment]
This should be good to go now.

**Comment by michaelpj** [comment]
Okay, let's give this a shot.

**Pull request merged**